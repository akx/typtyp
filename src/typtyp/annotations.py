import dataclasses


@dataclasses.dataclass(frozen=True)
class Comment:
    comment: str
