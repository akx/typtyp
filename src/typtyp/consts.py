import collections

COLLECTION_ORIGINS = (
    list,
    collections.deque,
    set,
    frozenset,
    collections.abc.Iterable,
    collections.abc.Iterator,
    collections.abc.Sequence,
)

MAPPING_ORIGINS = (
    dict,
    collections.defaultdict,
    collections.abc.Mapping,
    collections.abc.MutableMapping,
    collections.OrderedDict,
)
