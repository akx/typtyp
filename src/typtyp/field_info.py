import dataclasses
from typing import TypedDict


@dataclasses.dataclass(frozen=True)
class FieldInfo:
    name: str
    type: type
    doc: str | None = None


class FieldInfoDict(TypedDict, total=False):
    name: str
    type: type
    doc: str | None
