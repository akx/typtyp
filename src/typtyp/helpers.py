from typing import Hashable, Iterable, TypeVar

THashable = TypeVar("THashable", bound=Hashable)


def unique_in_order(values: Iterable[THashable]) -> Iterable[THashable]:
    seen = set()
    for v in values:
        if v in seen:
            continue
        seen.add(v)
        yield v
