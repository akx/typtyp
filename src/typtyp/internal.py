import dataclasses


@dataclasses.dataclass(frozen=True)
class FieldInfo:
    name: str
    type: type
    doc: str | None = None
