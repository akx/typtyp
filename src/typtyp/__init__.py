from typtyp.field_info import FieldInfo, FieldInfoDict
from typtyp.type_configuration import TypeConfiguration
from typtyp.world import World

__all__ = [
    "FieldInfo",
    "FieldInfoDict",
    "TypeConfiguration",
    "World",
]
