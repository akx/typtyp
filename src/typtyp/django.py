from __future__ import annotations

import datetime
import uuid
from decimal import Decimal
from typing import Annotated, Any, Iterable, Literal, Optional, Union

from django.db import models

from typtyp.annotations import Comment
from typtyp.internal import FieldInfo

STRINGLIKE_FIELDS = (
    models.CharField,
    models.EmailField,
    models.GenericIPAddressField,
    models.SlugField,
    models.TextField,
    models.URLField,
)

SIMPLE_FIELD_TYPES = {
    models.BinaryField: bytes,
    models.BooleanField: bool,
    models.DateField: datetime.date,
    models.DateTimeField: datetime.datetime,
    models.DecimalField: Decimal,
    models.DurationField: str,
    models.FileField: Annotated[str, Comment("file field")],  # TODO: Handle more specific types?
    models.FloatField: float,
    models.ImageField: Annotated[str, Comment("image field")],  # TODO: Handle more specific types?
    models.IntegerField: int,
    models.JSONField: Any,
    models.TimeField: datetime.time,
    models.UUIDField: uuid.UUID,
}


def get_referred_model_type(field: models.Field, model_type: type[models.Model]):
    if field.related_model == "self":
        return model_type  # self-referential relation
    return field.related_model


def get_django_field_type(field: models.Field, *, model_type: type[models.Model]) -> tuple[type, str | None] | None:
    if flatchoices := getattr(field, "flatchoices", None):
        return (Union[*(Literal[c[0]] for c in flatchoices)], None)
    if isinstance(field, STRINGLIKE_FIELDS):
        return (str, None)
    for field_type, typ in SIMPLE_FIELD_TYPES.items():
        if isinstance(field, field_type):
            return (typ, None)
    if isinstance(field, (models.ManyToManyRel, models.ManyToOneRel)):
        # Skip reverse relations for now.
        return None
    if isinstance(field, models.ManyToManyField):
        return (list[get_referred_model_type(field, model_type)], "many-to-many relation")
    if isinstance(field, models.OneToOneField):
        return (get_referred_model_type(field, model_type), "one-to-one relation")
    if isinstance(field, models.ForeignKey):
        return (get_referred_model_type(field, model_type), "foreign key relation")
    raise NotImplementedError(f"Unsupported Django field type ({type(field).__name__}): {field!r}")


def is_django_model(t) -> bool:
    """
    Infallibly checks if a type is a Django model. Never throws.
    """
    try:
        return issubclass(t, models.Model)
    except Exception:
        return False


def get_model_fields(model_type: type[models.Model]) -> Iterable[FieldInfo]:
    for field in sorted(model_type._meta.get_fields(include_hidden=True), key=lambda field: field.name):
        typ = get_django_field_type(field, model_type=model_type)
        if typ is None:
            continue
        typ, comment = typ
        if field.null:
            typ = Optional[typ]
        if comment:
            typ = Annotated[typ, Comment(comment=comment)]
        yield FieldInfo(field.name, typ, doc=str(field.help_text) if field.help_text else None)
