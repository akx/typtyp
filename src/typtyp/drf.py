from __future__ import annotations

import datetime
import inspect
import uuid
from decimal import Decimal
from typing import Annotated, Any, Iterable, Literal, Union

from rest_framework import fields, relations, serializers

from typtyp.annotations import Comment
from typtyp.field_info import FieldInfo

STRINGLIKE_FIELDS = (
    fields.CharField,
    fields.EmailField,
    fields.RegexField,
    fields.SlugField,
    fields.URLField,
)

SIMPLE_FIELD_TYPES = {
    fields.BooleanField: bool,
    fields.DateField: datetime.date,
    fields.DateTimeField: datetime.datetime,
    fields.DecimalField: Decimal,
    fields.DurationField: str,
    fields.FileField: Annotated[str, Comment("file field")],  # TODO: Handle more specific types?
    fields.FloatField: float,
    fields.ImageField: Annotated[str, Comment("image field")],  # TODO: Handle more specific types?
    fields.IntegerField: int,
    fields.JSONField: Any,
    fields.TimeField: datetime.time,
    fields.ReadOnlyField: Annotated[  # TODO: can we dig into these?
        Any,
        Comment(comment="read only field, type unknown"),
    ],
    fields.UUIDField: uuid.UUID,
    relations.HyperlinkedRelatedField: str,
    relations.PrimaryKeyRelatedField: Annotated[  # TODO: Handle more specific types?
        Any,
        Comment(comment="primary key related field"),
    ],
}


def is_drf_serializer(t) -> bool:
    """
    Infallibly checks if a type is a DRF serializer. Never throws.
    """
    try:
        return issubclass(t, serializers.Serializer)
    except Exception:  # pragma: no cover
        return False


def get_drf_field_type(field: fields.Field) -> type:
    if isinstance(field, serializers.DictField):
        return dict[Any, get_drf_field_type(field.child)]
    if isinstance(field, fields._UnvalidatedField):
        return Any
    if isinstance(field, fields.ListField):
        return list[get_drf_field_type(field.child)]
    if isinstance(field, serializers.Serializer):
        return field.__class__
    if isinstance(field, serializers.ListSerializer):
        return list[get_drf_field_type(field.child)]
    if isinstance(field, fields.ChoiceField):
        return Union[*(Literal[choice] for choice in field.choices)]
    if isinstance(field, serializers.SerializerMethodField):
        method = getattr(field.parent, field.method_name)
        rtype = inspect.get_annotations(method).get("return", Any)
        if isinstance(rtype, str):
            # If the return type is a string, assume it's a forward reference
            return Annotated[
                Any,
                Comment(comment=f"forward reference: {rtype}"),
            ]
        return rtype
    if isinstance(field, STRINGLIKE_FIELDS):
        return str
    for field_type, typ in SIMPLE_FIELD_TYPES.items():
        if isinstance(field, field_type):
            return typ
    raise NotImplementedError(f"Unsupported DRF field type ({type(field).__name__}): {field!r}")  # pragma: no cover


def get_serializer_fields(ser_type: type[serializers.Serializer]) -> Iterable[FieldInfo]:
    context = {"request": None}
    for name, field in sorted(ser_type(context=context).fields.items(), key=lambda item: item[0]):
        yield FieldInfo(
            name=name,
            type=get_drf_field_type(field),
            doc=str(field.help_text) if field.help_text else None,
        )
