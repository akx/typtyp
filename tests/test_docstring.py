import typtyp
from tests.helpers import check_with_tsc


def test_no_builtin_docstring(snapshot):
    w = typtyp.World()
    w.add(list[int], name="ListOfInts")
    code = w.get_typescript()
    assert check_with_tsc(code)
    assert code == snapshot
