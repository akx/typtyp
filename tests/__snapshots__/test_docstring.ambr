# serializer version: 1
# name: test_no_builtin_docstring
  '''
  export type ListOfInts = (number)[]
  
  '''
# ---
